### PR TITLE
Compile tests action runs

### PR DESCRIPTION
**Pull request by grantcooksey**
This patch is a step towards getting the CI working.
* Some sketches are missing, these were commented out
* The matrix exclusions weren't respecting the null soft-serial option so the test matrix is fully specified
* Pretty much all of the sketches are failing to compile but they are now running

### CODE REVIEW

**Comment by SpenceKonde** [comment]
hell yeah lets check this out

**Comment by grantcooksey** [comment]
@SpenceKonde Looks like a couple classes of errors:
* compilation bugs - Some of these I'm not totally sure about but I'm trying to work through the logic in the compile-examples github action to make sure I didn't miss something
* missing fqbn - From what I understand, it looks like the action is loading the latest ATTinyCore package from the `http://drazzy.com/package_drazzy.com_index.json` index, which looks like version 1.5.2. I pulled that version and poked around `boards.txt` and didn't find some of the boards listed in the matrix. `ATTinyCore:avr:attiny26` is one of them for reference.  Any ideas here?

**Comment by SpenceKonde** [comment]
Oh all the FQBNs are different in 2.0.0, or a great many of them are/will be :-/ 

Because I added a shitload of features and options that add new boards.

I fully expect to turn up tons of compile errors the first time the CI is run on 2.0.0 codebase with correct FQBNs. That's why I'm unwilling to release without it. 

We also are limited to 100 jobs as I understand (I think this core needs several times that :-/ ). You can see what I did over on megaTinyCore to deal with that -  I have one job for each chip, and then run compile examples with like 8 fqbn's each to get coverage of everything I know will go through different codepaths

**Comment by SpenceKonde** [comment]
But are you implyingthat this compile examples code can't be used on an "all new" core, because it needs to download and install an older version to get information it needs.... ?

**Comment by SpenceKonde** [comment]
That means that there would have to be a released version that everyone's IDE would tell them was an upgrade.... but which I knew damned well would be totally hosed doesn't it? (because I am confident that the CI will turn up probably 20+ unknown bugs the first time it's run)

**Comment by grantcooksey** [comment]
Ah I see, this makes sense. We can't use the package index because we need to test against the unreleased version within the current branch.

I can try out using the [`local-path` option](https://github.com/arduino/compile-sketches#local-path).

> We also are limited to 100 jobs as I understand

Github's docs are saying it's a bit higher, 256 max matrix jobs. Do you think that's enough room to get this started? 
I'm wrapping my head around the compile-examples action logic but it probably wouldn't be too hard to either fork that repo or create a local script that adds a wrapper to consolidate the matrix a bit. It would be nice if this could be a separate effort though.

**Comment by SpenceKonde** [comment]
if needed for initial testing, a json file containing a known bad "release" can be made. until we have something plausibile enough to put on the normal public release file

**Comment by grantcooksey** [comment]
Ok, no need to make a bad release.

The action's platform installation has this:
```
platforms: |
            # Overwrite the ATTinyCore release version with the platform from the local path
            - source-path: avr
              name: ${{ env.platform-name }}
            # Install ATTinyCore via Boards Manager for the toolchain
            - source-url: http://drazzy.com/package_drazzy.com_index.json
              name: ${{ env.platform-name }}
```

The `platform` override logic is actually working but it's a little confusing because `arduino-cli` reports the older overridden version rather than the actual version. You can see in the CI logs that the platform is getting reported as 1.5.2 even though the actual version is 2.0.0-dev.

Looks like the compile-sketches script swaps out the directory so this is working as expected.
```
root@f03bb7ff6607:~# cat .arduino15/packages/ATTinyCore/hardware/avr/1.5.2/platform.txt | grep '^versionnum'
versionnum.major=2
versionnum.minor=0
versionnum.patch=0
versionnum.postfix=-dev
versionnum.released=0
root@f03bb7ff6607:~# ./bin/arduino-cli core list
ID             Installed Latest Name      
ATTinyCore:avr 1.5.2     1.5.2  ATTinyCore

```

Other issues that I've seen:
* Some FQBNs are not resolving. For example: `ATTinyCore:avr:attiny26,clock=internal_8m,softserial=enable` fails but `ATTinyCore:avr:attiny26` works, any ideas?
* `avr/extras/ci/test_sketches_2k/basic_test/basic_test.ino` uses the macro `GPIO0` but I can't find that symbol definition in this project. I did a very light search and couldn't find a common pin name that we could use here, do you know of one? Otherwise, that test might need to tweak that test to support swapping the pin out based on the board.
* The SPI library also has a number of missing symbols. For example, the pin definitions in the current HEAD don't define `USI_SCK_PORT` anywhere but I checked out the 1.5.2 tag and I see that symbol defined in all the `pins_arduino.h` variants. I guess we just need to port those over?

**Comment by SpenceKonde** [comment]
> Ok, no need to make a bad release.
> 
> The action's platform installation has this:
> 
> ```
> platforms: |
>             # Overwrite the ATTinyCore release version with the platform from the local path
>             - source-path: avr
>               name: ${{ env.platform-name }}
>             # Install ATTinyCore via Boards Manager for the toolchain
>             - source-url: http://drazzy.com/package_drazzy.com_index.json
>               name: ${{ env.platform-name }}
> ```
> 
> The `platform` override logic is actually working but it's a little confusing because `arduino-cli` reports the older overridden version rather than the actual version. You can see in the CI logs that the platform is getting reported as 1.5.2 even though the actual version is 2.0.0-dev.
> 
> Looks like the compile-sketches script swaps out the directory so this is working as expected.
> 
> ```
> root@f03bb7ff6607:~# cat .arduino15/packages/ATTinyCore/hardware/avr/1.5.2/platform.txt | grep '^versionnum'
> versionnum.major=2
> versionnum.minor=0
> versionnum.patch=0
> versionnum.postfix=-dev
> versionnum.released=0
> root@f03bb7ff6607:~# ./bin/arduino-cli core list
> ID             Installed Latest Name      
> ATTinyCore:avr 1.5.2     1.5.2  ATTinyCore
> ```
> 
sigh *smh* of course, it sees 1.5.2 installed because we've disguised 2.0.0 in the 1.5.2 folders and board manager knows it's only ever installed a 1.5.2. 

 
> * Some FQBNs are not resolving. For example: `ATTinyCore:avr:attiny26,clock=internal_8m,softserial=enable` fails but `ATTinyCore:avr:attiny26` works, any ideas?

What the fuck?

> * `avr/extras/ci/test_sketches_2k/basic_test/basic_test.ino` uses the macro `GPIO0` but I can't find that symbol definition in this project. I did a very light search and couldn't find a common pin name that we could use here, do you know of one? Otherwise, that test might need to tweak that test to support swapping the pin out based on the board.

Supplied by the microchip headers. These have been referred to as GPIOn, GPIORn, and now GPR.GPRn. It's handy because it's a free-to-use volatile register that has minimal cost to access - so it forces the compiler to not optimize away reads or writes,

All devices have the pin LED_BUILTIN defined, precisely to make the examples run. 

> * The SPI library also has a number of missing symbols. For example, the pin definitions in the current HEAD don't define `USI_SCK_PORT` anywhere but I checked out the 1.5.2 tag and I see that symbol defined in all the `pins_arduino.h` variants. I guess we just need to port those over?

Those get generated Arduino.h after importing pins_arduino.h (a lot of stuff does)

**Comment by jvasileff** [comment]
> * Some FQBNs are not resolving. For example: `ATTinyCore:avr:attiny26,clock=internal_8m,softserial=enable` fails but `ATTinyCore:avr:attiny26` works, any ideas?

The FQBN `ATTinyCore:avr:attiny26,clock=internal_8m,softserial=enable` is invalid - the delimiter after `attiny26` must be `:`, not `,`. This adds complexity to forming the FQBN since some devices already have an initial parameter, e.g. `attinyx61:chip=861`, for which `,` would be valid for the delimiter before `clock=...`. See [this comment](https://github.com/SpenceKonde/ATTinyCore/issues/667#issuecomment-1251651116) for the syntax.

-----------

Before I saw this PR, I worked through code to handle all of the complexities of using a matrix to form valid FQBNs, but the result, even though it worked, is too complex to be maintainable IMO, especially when considering the `attinyx61` with non-standard `softserial` values like `enable_AIN1`.

So rather than using the matrix functionality to form a Cartesian product of options, and then tweaking with includes and excludes, *and* dealing with delimiter problems, I recommend just forming the FQBNs manually, as in:

```
      matrix:
        device:
          - name: attiny84
            available-flash-kB: 8
            fqbn: ATTinyCore:avr:attinyx4:chip=84,clock=internal_8m,softserial=enable
```

This is close to what you have, but also punts on programmatically forming the FQBNs.

Then, perhaps move `available-flash-kB: 8` back to an `include:` section to avoid repeating for each unique fqbn for each chip.

**Comment by henrygab** [comment]
<details><summary>No response ... moved on</summary><P/>
Hi @SpenceKonde,

I actually was just looking to enable CI to validate one of my libraries with v2.

I'd like to make a significant recommendation, which will save you significant headache, but is fairly simple:

### Change the `vendor` portion of FQBN for v2.x 

Specifically, don't continue using `ATTinyCore`, but rather change it (to anything else ... maybe `ATTinyCore2`).  This will avoid confusion when the `architecture` and `board` portions match.

Example of why:

* Simplifies Action matrix use ... by allowing unique listings for ATTinyCore v1 and ATTinyCore v2
    * See, for example, https://github.com/SimpleHacks/QDEC/blob/072930b46a9169111119ba201101067ca658c763/.github/workflows/arduino.yml#L118-L145
    * I build a bunch of different platforms, with many, many package_index sources.
    * By listing the platforms individually, I only need the FQBN strings listed in the 
* Allows folks to install v1.x side-by-side with v2.x
* Avoid all the problems associated with "hiding" the known earlier versions during CI

**_YES_** -- the Arduino/compile-sketches action is painfully brute-force.  All your earlier comments are spot-on ... it would be *SO* much more efficient if the tools and other stuff gets cached.   I also know that's a possibility, but don't have the experience needed to make it happen.  :(

**_YES_** -- writing a better action would greatly reduce CI build times.  Something that could directly parse boards.txt for a given board, AFTER all the tooling was setup, and build a binary for every combination of options.   Again, I know it's possible, but I don't have the experience needed to make it happen either. :(

Sadly, the Arduino/compile-sketches Action appears fragile to me.  I have to re-run failed builds regularly, and I've only gotten them working in the last week or so.  I think the underlying fragility relates to too many HTTP requests for the tools (anti-flood filters).  Thus, if you group things more rationally than I did (I brute force one job per FQBN .. and each one downloads all the tools, etc...), that likely would not be as significant an issue.  At least the minutes are free....

Maybe this weekend, I can fork this and play around some ... see if I can make the above vendor name change, and get a much smaller set of the CI compilation-focused checks working.


</details>

Bummer ... Please presume that this is no longer on my radar, as other things now are taking my time.

**Comment by henrygab** [comment]
<details><summary>No response ... moved on</summary><P/>
FYI ... I'm starting by working to get caching of the `arduino-cli` (with all the tools pre-installed) working.  That should shave some build & compile time.   I've also started poking my fingers into the script you use to generate `boards.txt`.

The more I see the complexity here, the more I think you're going to want to run a script (similar to what you do for boards.txt ... maybe the same script) to generate the CI build workflow YAML files.  _(Else, it's just asking for typos and maintenance headaches.)_

@SpenceKonde -- when this starts to get closer to working well, the following will be needed from you:

Conceptually, presume that each board gets its own workflow file generated.  Each board has a large suite of possible menu items exposed in boards.txt ...
1. If five menu options, and each menu item has 4 options, do you want to verify compilation for all 1024 permutations?
1. If not, which permutations are most important to build for all boards?
1. Which permutations need coverage, but maybe not for all boards?
1. Are there "classes" of boards, where you want 100% coverage of the permutations in that "class" of board (but OK to only build the most important on the individual members of the class)?
1. Are there other groupings that make sense to you?

Thanks!
</details>

Bummer ... Please presume that this is no longer on my radar, as other things now are taking my time.

**Comment by henrygab** [comment]
@SpenceKonde 

<details><summary>No response ... moved on</summary><P/>

Had some more time, and wanted to give you an update, and ask some questions:

* I have successfully gotten caching of the `arduino-cli` (with tools pre-installed) working.
  * This reduces job time, as installation only needs to occur once
  * It also prevents random compilation failures caused by external sites rate-limiting (temporarily blocking) requests ... a fairly significant benefit
* I'm currently starting at v1.5.2 ... which means all the FQBN options are wrong for v2.x ... but it at least shows how to use the CI builds.

If you want to see all my mistakes, check out my [messy](https://github.com/henrygab/ATTinyCore/commits/v2.0.0-devX) commits... .

Given the complexity of the repo, it seems creating final workflows by hand would be ... error prone.  The good news is that it appears possible to generate the workflow files programmatically, similar to how you generate boards.txt.

* allows changing how the workflows are split ... e.g. to avoid job count limits 
* allows creating workflows to build on a schedule, e.g. to get 100% boards.txt option coverage every N days

## Questions for you....

There are too many combinations for even a single board to realistically expect to test all combinations.

1. Can you help me understand what combinations of FQBN settings are most critical to ensure get CI builds going?
2. Which options / permutations need coverage, but would be OK with only one board using the option in CI builds?
3. Which options / permutations do you want to build *all* boards against (e.g., default settings?)
4. Can you divide the chips into "classes", such that you want builds to cover 100% of the option permutations in that "class"?
5. etc.

Basically, if you could define the spectrum of what gets built, and how often... what would that be?

Alternatively, if no longer of interest to get this all working happily, that would also be good to know....

</details>

Bummer ... Please presume that this is no longer on my radar, as other things now are taking my time.

**Comment by SpenceKonde** [comment]
Well, I've finally had aa chance to come back to this. I keep getting pulled off of this to work on more urgent things. DxCore needed to get several new parts added, and one piece of plumbing turned into a multi-phase nightmare. Then a client needed custom design work urgently, then critical bugs were found in DxCore. So I've sort of been bouncing between disasters trying to make things work and keep up with the pace of modern AVR development
